test/syscalls/linux/alloc_fd_benchmark.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) 2019 Ant Financial Services Group.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sys/mman.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <random>
+
+#include "benchmark/benchmark.h"
+#include "gtest/gtest.h"
+
+namespace gvisor {
+namespace testing {
+
+namespace {
+
+const int DEFAULT_FILE_NUM = 10000;
+const char *filename = "memfd_test";
+
+// Filename used by memfd_create actually doesn't matter in this test case.
+// "You can create multiple files with the same name
+// and they will all be separate, independent files."
+// ref: https://dvdhrm.wordpress.com/2014/06/10/memfd_create2/
+
+void BM_open_serial(benchmark::State &state) {
+  auto fd_table = std::vector<int>(DEFAULT_FILE_NUM);
+  for (auto _ : state) {
+    for (int i = 0; i < DEFAULT_FILE_NUM; ++i) {
+      fd_table[i] = memfd_create(filename, MFD_CLOEXEC);
```

Comment:
can we use dup instead of memfd_create?
```
 void BM_open_serial(benchmark::State &state) {
   auto fd_table = std::vector<int>(DEFAULT_FILE_NUM);
+  FileDescriptor fd = ASSERT_NO_ERRNO_AND_VALUE(Open("/dev/null", O_RDONLY));
   for (auto _ : state) {
     for (int i = 0; i < DEFAULT_FILE_NUM; ++i) {
-      fd_table[i] = memfd_create(filename, MFD_CLOEXEC);
+      EXPECT_THAT(fd_table[i] = dup(fd.get()), SyscallSucceeds());
     }
     for (int i = 0; i < DEFAULT_FILE_NUM; ++i) {
-      close(fd_table[i]);
+      EXPECT_THAT(close(fd_table[i]), SyscallSucceeds());
     }
   }
 }
```

Reply:
:ok_hand:

Reply:
One more thing I want to discuss, in order to use `FileDescriptor`, `FileDescriptor` and so on,
I have to add lines in test/syscall/linux/BUILD:
```
deps = [
+    "//test/util:file_descriptor",
+    "//test/util:temp_path",
+    "//test/util:test_main",
+    "//test/util:test_util",
    "@com_google_benchmark//:benchmark_main",
    "@com_google_googletest//:gtest",
],
```
and in test/syscall/linux/alloc_fd_benchmark.cc:
```c
#include "test/util/test_util.h"
#include "test/util/file_descriptor.h"
```
But changes above will make `bazel run //test/syscalls/linux:alloc_fd_benchmark` output become like this:
```
  bazel-bin/test/syscalls/linux/alloc_fd_benchmark
INFO: Elapsed time: 1.484s, Critical Path: 1.32s
INFO: 2 processes: 2 linux-sandbox.
INFO: Build completed successfully, 3 total actions
INFO: Build completed successfully, 3 total actions
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```
What I want is benchmark not functional test (Function test already exists):
```
2019-09-29 07:09:41
Running /root/alloc_fd_benchmark
Run on (96 X 2699.89 MHz CPU s)
***WARNING*** Library was built as DEBUG. Timings may be affected.
----------------------------------------------------------
Benchmark                   Time           CPU Iterations
----------------------------------------------------------
BM_open_serial       36310447 ns   36315789 ns         19
BM_open_zigzag       20871677 ns   20625000 ns         32
BM_open_one_by_one   36151259 ns   35789474 ns         19
BM_open_random       18542792 ns   18510638 ns         47
```

Any ideas?

Reply:
I think you need to specify "-benchmarks all" options:
$ test/syscalls/linux/alloc_fd_benchmark -benchmarks

Here is another problem. Our syscall_test_runner doesn't know about this -benchmark option too. Actually, we have a set of benchmarks for gvisor and an infrastructure to run them, but unfortunate this part isn't opensourced yet.

So you have to options:
1. You can teach syscall_test_runner to run benchmarks.
2. Or I can place your benchmark near with other gvisor benchmarks and then one day they all will be open-sourced.

Reply:
Well, seems like "Choice 2" is more systematism. I choose 2. I'll push newer patch today later.
(Sorry for reply late, I was on festival vacation.)